src/graphql/outlier.rs

```diff
@@ -10,19 +10,117 @@ use async_graphql::{
     connection::{query, Connection, Edge, EmptyFields},
     types::ID,
     ComplexObject, Context, InputObject, Object, ObjectType, OutputType, Result, SimpleObject,
+    Subscription,
 };
 use bincode::Options;
 use chrono::{offset::LocalResult, DateTime, NaiveDateTime, TimeZone, Utc};
 use data_encoding::BASE64;
+use futures::channel::mpsc::{unbounded, UnboundedSender};
+use futures_util::stream::Stream;
 use num_traits::ToPrimitive;
-use review_database::{types::FromKeyValue, Database, Direction, IterableMap};
+use review_database::{types::FromKeyValue, Database, Direction, IterableMap, Store};
 use serde::Deserialize;
 use serde::Serialize;
-use std::cmp;
+use std::{cmp, collections::HashMap, sync::Arc};
+use tokio::{sync::RwLock, time};
+use tracing::error;
 
 pub const TIMESTAMP_SIZE: usize = 8;
 const DEFAULT_OUTLIER_SIZE: usize = 50;
 const DISTANCE_EPSILON: f64 = 0.1;
+const DEFAULT_RANKED_OUTLIER_FETCH_TIME: u64 = 60;
+const MAX_MODEL_LIST_SIZE: usize = 100;
+
+#[derive(Default)]
+pub(super) struct OutlierStream;
+
+#[Subscription]
+impl OutlierStream {
+    #[graphql(guard = "RoleGuard::new(Role::SystemAdministrator)
+        .or(RoleGuard::new(Role::SecurityAdministrator))
+        .or(RoleGuard::new(Role::SecurityManager))
+        .or(RoleGuard::new(Role::SecurityMonitor))")]
+    async fn ranked_outlier_stream(
+        &self,
+        ctx: &Context<'_>,
+        start: DateTime<Utc>,
+        fetch_interval: Option<u64>,
+    ) -> Result<impl Stream<Item = RankedOutlier>> {
+        let store = ctx.data::<Arc<RwLock<Store>>>()?.clone();
+        let db = ctx.data::<Database>()?.clone();
+
+        let fetch_time = if let Some(fetch_time) = fetch_interval {
+            fetch_time
+        } else {
+            DEFAULT_RANKED_OUTLIER_FETCH_TIME
+        };
```

Comment:
Is there a reason you didn't use `unwrap_or`?

Reply:
I have updated the code for use `unwrap_or`.